src/Glacier/MultipartUploader.php

```diff
@@ -0,0 +1,278 @@
+<?php
+namespace Aws\Glacier;
+
+use Aws\CommandInterface;
+use Aws\HashingStream;
+use Aws\Multipart\AbstractUploader;
+use Aws\Multipart\UploadState;
+use Aws\PhpHash;
+use Aws\ResultInterface;
+use GuzzleHttp\Psr7;
+use Psr\Http\Message\StreamableInterface as Stream;
+
+/**
+ * Encapsulates the execution of a multipart upload to Glacier.
+ */
+class MultipartUploader extends AbstractUploader
+{
+    const PART_MIN_SIZE = 1048576;
+
+    private static $validPartSizes = [
+        1048576,    //   1 MB
+        2097152,    //   2 MB
+        4194304,    //   4 MB
+        8388608,    //   8 MB
+        16777216,   //  16 MB
+        33554432,   //  32 MB
+        67108864,   //  64 MB
+        134217728,  // 128 MB
+        268435456,  // 256 MB
+        536870912,  // 512 MB
+        1073741824, //   1 GB
+        2147483648, //   2 GB
+        4294967296, //   4 GB
+    ];
+
+    /**
+     * Creates an UploadState object for a multipart upload by querying the
+     * service for the specified upload's information.
+     *
+     * @param GlacierClient $client    GlacierClient object to use.
+     * @param string        $vaultName Vault name for the multipart upload.
+     * @param string        $uploadId  Upload ID for the multipart upload.
+     * @param string        $accountId Account ID for the multipart upload.
+     *
+     * @return UploadState
+     */
+    public static function getStateFromService(
+        GlacierClient $client,
+        $vaultName,
+        $uploadId,
+        $accountId = '-'
+    ) {
+        $state = new UploadState([
+            'accountId' => $accountId,
+            'vaultName' => $vaultName,
+            'uploadId'  => $uploadId,
+        ]);
+
+        foreach ($client->getPaginator('ListParts', $state->getId()) as $result) {
+            // Get the part size from the first part in the first result.
+            if (!$state->getPartSize()) {
+                $state->setPartSize($result['PartSizeInBytes']);
+            }
+            // Mark all the parts returned by ListParts as uploaded.
+            foreach ($result['Parts'] as $part) {
+                list($rangeIndex, $rangeSize) = self::parseRange(
+                    $part['RangeInBytes'],
+                    $state->getPartSize()
+                );
+                $state->markPartAsUploaded($rangeIndex, [
+                    'size'     => $rangeSize,
+                    'checksum' => $part['SHA256TreeHash'],
+                ]);
+            }
+        }
+
+        $state->setStatus(UploadState::INITIATED);
+
+        return $state;
+    }
+
+    /**
+     * Creates a multipart upload for a Glacier archive.
+     *
+     * The valid configuration options are as follows:
+     *
+     * - account_id: (string, required, default=string('-')) Account ID for the
```

Comment:
It's needed so it can be overridden if the archive is for an account other than the one performing the operations, but I updated the docblock to remove "required" and added a clarification.